src/cryptonote_core/service_node_list.cpp

```diff
@@ -1814,101 +1706,97 @@ namespace service_nodes
+    data_for_serialization new_data_in = {};
+    {
+      old_data_members_for_serialization old_data_in = {};
+      bool old_data                                  = false;
+      bool new_data                                  = ::serialization::serialize(ba, new_data_in);
+      if (!new_data)
+      {
+        std::stringstream ss_old;
+        ss_old << blob;
+        binary_archive<false> ba_old(ss_old);
+        old_data = ::serialization::serialize(ba_old, old_data_in);
+      }
+
+      CHECK_AND_ASSERT_MES(old_data || new_data, false, "Failed to parse service node data from blob");
+      if (old_data)
+      {
+        new_data_in.states.push_back({});
+        new_data_in.quorum_states   = std::move(old_data_in.quorum_states);
+        state_serialized &new_state = new_data_in.states.back();
 
-    for (const auto& states : data_in.quorum_states)
+        new_data_in.version           = old_data_in.version;
+        new_state.height              = old_data_in.height;
+        new_state.infos               = std::move(old_data_in.infos);
+        new_state.key_image_blacklist = std::move(old_data_in.key_image_blacklist);
+      }
+    }
+
+    if (new_data_in.states.empty() || new_data_in.quorum_states.empty())
+      return false;
+
+    for (auto& quorum_serialized : new_data_in.quorum_states)
     {
-      testing_quorum const &obligations = states.quorums[static_cast<uint8_t>(quorum_type::obligations)];
-      m_transient_state.quorum_states[states.height].obligations = std::make_shared<testing_quorum>(obligations);
+      testing_quorum &obligations = quorum_serialized.quorums[static_cast<uint8_t>(quorum_type::obligations)];
+      uint64_t quorum_height      = quorum_serialized.height;
+      m_quorum_states[quorum_height].obligations = std::make_shared<testing_quorum>(std::move(obligations));
 
-      testing_quorum const &checkpointing = states.quorums[static_cast<uint8_t>(quorum_type::checkpointing)];
+      testing_quorum &checkpointing = quorum_serialized.quorums[static_cast<uint8_t>(quorum_type::checkpointing)];
       // Don't load any checkpoints that shouldn't exist (see the comment in generate_quorums as to
       // why the `+BUFFER` term is here).
-      if ((states.height + REORG_SAFETY_BUFFER_BLOCKS_POST_HF12) % CHECKPOINT_INTERVAL == 0)
-          m_transient_state.quorum_states[states.height].checkpointing = std::make_shared<testing_quorum>(checkpointing);
+      if ((quorum_height + REORG_SAFETY_BUFFER_BLOCKS_POST_HF12) % CHECKPOINT_INTERVAL == 0)
+        m_quorum_states[quorum_height].checkpointing = std::make_shared<testing_quorum>(std::move(checkpointing));
     }
 
-    for (const auto& info : data_in.infos)
+    bool first_time = true;
+    for (state_serialized &state_serialized : new_data_in.states)
     {
-      m_transient_state.service_nodes_infos[info.pubkey] = info.info;
-    }
+      if (!first_time) m_state_history.push_back({});
```

Comment:
It's not clear to me why the new `m_state_history` element should not be added for the first element here; I don't see where it gets created (maybe I'm just missing it -- but if it isn't guaranteed to be there then the following `.back()` seems wrong).

Reply:
Should be addressed by not using the pointer method.